src/main/java/com/researchspace/service/inventory/impl/InventoryApiManagerImpl.java

```diff
@@ -236,77 +236,88 @@ public ApiInventorySearchResult convertToApiInventorySearchResult(
     return apiSearchResult;
   }
 
-  FileProperty generateInventoryFilePropertyFromBase64Image(
-      User user, String imageName, String base64Image, boolean thumbnail) throws IOException {
+  FileProperty saveImageFile(User user, String base64Image) throws IOException {
+    String imageExtension = ImageUtils.getExtensionFromBase64DataImage(base64Image);
+    byte[] imageBytes = ImageUtils.getImageBytesFromBase64DataImage(base64Image);
+    InputStream imageIS = new ByteArrayInputStream(imageBytes);
+    String contentsHash = CryptoUtils.hashWithSha256inHex(base64Image);
+    String filename = String.format("%s.%s", contentsHash, imageExtension);
+
+    return saveOrRetrieveImage(user, filename, imageIS, contentsHash);
+  }
 
+  FileProperty saveThumbnailImageFile(User user, String base64Image) throws IOException {
     String imageExtension = ImageUtils.getExtensionFromBase64DataImage(base64Image);
     byte[] imageBytes = ImageUtils.getImageBytesFromBase64DataImage(base64Image);
 
-    InputStream imageIS;
-    if (thumbnail) {
-      imageIS = createThumbnailFromImageBytes(imageBytes, imageExtension);
-      imageName += "_thumbnail";
-    } else {
-      imageIS = new ByteArrayInputStream(imageBytes);
+    InputStream thumbnailForHash = new ByteArrayInputStream(imageBytes);
+    String contentsHash =
+        CryptoUtils.hashWithSha256inHex(Arrays.toString(thumbnailForHash.readAllBytes()));
+    String filename = String.format("%s.%s", contentsHash, imageExtension);
+
+    try (InputStream imageIS = createThumbnailFromImageBytes(imageBytes, imageExtension)) {
+      return saveOrRetrieveImage(user, filename, imageIS, contentsHash);
     }
+  }
 
-    String fileName = imageName + "." + imageExtension;
-    return inventoryFileApiManager.generateInventoryFileProperty(user, fileName, imageIS);
+  /*
+  The same FileProperty (and therefore the same file on disk) can belong to many InventoryRecords.
+  Checks if a FileProperty already exists for the given user and hash of the contents of the image
+  and returns that if so. Otherwise, generates a new FileProperty.
+   */
+  private FileProperty saveOrRetrieveImage(
+      User user, String fileName, InputStream imageIs, String contentsHash) throws IOException {
+    Optional<FileProperty> existingFile =
+        getExistingFilePropertyForImage(contentsHash, user.getUsername());
+    if (existingFile.isPresent()) {
+      return existingFile.get();
+    } else {
+      return inventoryFileApiManager.generateInventoryFileProperty(
```

Comment:
Is this actually saving? would it be `saveNewInventoryFileProperty` a better method name?

Reply:
Yeah it also saves that actual file to the filestore eventually, as well as the `FileProperty` metadata. 

That's why i renamed the enclosing method from `generateInventoryFilePropertyFromBase64Image` to `saveOrRetrieveImage` because I was confused about where the file itself actually got saved.

I've renamed this method as well now, good shout.